File: crc_checksum.c
#include<stdio.h>
#include<string.h>
#include<stdlib.h>

#define N strlen(g)

char t[28], cs[28], g[] = "1101";
int a, e, c;

// Function to perform XOR operation
void xorFunction() {
    for (c = 1; c < N; c++)
        cs[c] = ((cs[c] == g[c]) ? '0' : '1');
}

// Function to calculate CRC (Cyclic Redundancy Check)
void crc() {
    for (e = 0; e < N; e++)
        cs[e] = t[e];
    do {
        if (cs[0] == '1')
            xorFunction();
        for (c = 0; c < N - 1; c++)
            cs[c] = cs[c + 1];
        cs[c] = t[e++];
    } while (e <= a + N - 1);
}

// Function to calculate the checksum
char* findChecksum(char* sentMessage, int k) {
    int i;
    int len = strlen(sentMessage);
    char* checksum = (char*)malloc(k + 1);
    for (i = 0; i < len; i += k) {
        char subMessage[k + 1];
        strncpy(subMessage, sentMessage + i, k);
        subMessage[k] = '\0';
        unsigned int sum = strtoul(subMessage, NULL, 2);
        sum = (sum + 1) % (1 << k);
        sprintf(checksum, "%s%0*u", checksum, k, sum);
    }
    return checksum;
}

// Function to check the receiver's checksum
char* checkReceiverChecksum(char* receivedMessage, int k, char* checksum) {
    int i;
    int len = strlen(receivedMessage);
    char* receiverSum = (char*)malloc(k + 1);
    for (i = 0; i < len; i += k) {
        char subMessage[k + 1];
        strncpy(subMessage, receivedMessage + i, k);
        subMessage[k] = '\0';
        unsigned int sum = strtoul(subMessage, NULL, 2);
        sum = (sum + strtoul(checksum, NULL, 2)) % (1 << k);
        sprintf(receiverSum, "%s%0*u", receiverSum, k, sum);
    }
    char* receiverChecksum = (char*)malloc(k + 1);
    for (i = 0; i < k; i++) {
        receiverChecksum[i] = (receiverSum[i] == '1') ? '0' : '1';
    }
    receiverChecksum[k] = '\0';
    return receiverChecksum;
}

// Function to perform the CRC calculation
void performCRC() {
    printf("Enter data: ");
    scanf("%s", t);
    printf("----------------------------------------\n");
    printf("Generating polynomial: %s\n", g);
    a = strlen(t);
    for (e = a; e < a + N - 1; e++)
        t[e] = '0';
    printf("----------------------------------------\n");
    crc();
    printf("CRC: %s\n", cs);
    for (e = a; e < a + N - 1; e++)
        t[e] = cs[e - a];
    printf("----------------------------------------\n");
    printf("Final data to be sent: %s\n", t);
    printf("----------------------------------------\n");
}

// Function to perform the checksum calculation
void performChecksum() {
    char sentMessage[100];
    int k;
    printf("Enter data: ");
    scanf("%s", sentMessage);
    printf("Enter the number of bits for checksum (k): ");
    scanf("%d", &k);
    printf("----------------------------------------\n");
    char* checksum = findChecksum(sentMessage, k);
    printf("Sender side checksum: %s\n", checksum);
    char receivedMessage[100];
    printf("Enter received data: ");
    scanf("%s", receivedMessage);
    char* receiverChecksum = checkReceiverChecksum(receivedMessage, k, checksum);
    printf("Receiver side checksum: %s\n", receiverChecksum);
    printf("----------------------------------------\n");
    char* finalSum = (char*)malloc(k + 1);
    int i;
    for (i = 0; i < k; i++) {
        finalSum[i] = ((checksum[i] - '0') + (receiverChecksum[i] - '0')) % 2 + '0';
    }
    finalSum[k] = '\0';
    if (atoi(finalSum) == 0) {
        printf("Receiver Checksum = 0.\n");
        printf("Data Accepted\n");
    }
    else {
        printf("Receiver Checksum != 0.\n");
        printf("Error\n");
    }
    printf("----------------------------------------\n");
    free(checksum);
    free(receiverChecksum);
    free(finalSum);
}

int main() {
    int choice;
    while (1) {
        printf("Menu:\n");
        printf("1. Calculate CRC\n");
        printf("2. Calculate Checksum\n");
        printf("3. Exit\n");
        printf("Enter your choice: ");
        scanf("%d", &choice);
        fflush(stdin);

        switch (choice) {
        case 1:
            performCRC();
            break;

        case 2:
            performChecksum();
            break;

        case 3:
            exit(0);

        default:
            printf("Invalid choice! Please try again.\n\n");
            break;
        }
    }
    return 0;
}

File: crc_checksum.py
def crc(data, divisor):
    # Append zeros to the data to match the length of the divisor
    data += "0" * (len(divisor) - 1)

    # Convert data and divisor to lists for easier manipulation
    data = list(data)
    divisor = list(divisor)

    # Perform division
    for i in range(len(data) - len(divisor) + 1):
        if data[i] == "1":
            for j in range(len(divisor)):
                data[i + j] = str(int(data[i + j]) ^ int(divisor[j]))

    # Return the remainder (CRC value)
    return "".join(data[-(len(divisor)-1):])


def crc_check(data, divisor, crc_value):
    data += str(crc_value)
    data = list(data)
    divisor = list(divisor)

    # Perform division
    for i in range(len(data) - len(divisor) + 1):
        if data[i] == "1":
            for j in range(len(divisor)):
                data[i + j] = str(int(data[i + j]) ^ int(divisor[j]))

    # Return the remainder (CRC value)
    return "".join(data[-(len(divisor)-1):])


def calculate_checksum(data):
    checksum = 0

    for value in data:
        checksum += int(value, 2)  # Convert binary string to decimal

    checksum_bits = 8 if len(data[0]) <= 8 else 16

    if checksum_bits == 8:
        complement = ~checksum
        complement %= 256  # Take the modulo 256 to get an 8-bit checksum value
        return bin(complement)[2:].zfill(8)  # Convert checksum to 8-bit binary string
    else:
        complement = ~checksum
        complement %= 65536  # Take the modulo 65536 to get a 16-bit checksum value
        return bin(complement)[2:].zfill(16)  # Convert checksum to 16-bit binary string


def input_check(data):
    for i in data:
        if i != '0' and i != '1':
            return False
    return True


# Menu-driven program
while True:
    print("Menu:")
    print("1. CRC and CRC check")
    print("2. Checksum")
    print("3. Exit")

    choice = input("Enter your choice: ")

    if choice == "1":
        data_input = input("Enter data: ")
        divisor_input = input("Enter divisor: ")

        if not input_check(data_input) or not input_check(divisor_input):
            print("Invalid input! Please enter binary strings.")
            continue

        crc_value = crc(data_input, divisor_input)
        print("CRC value:", crc_value)

        check_input = input("Enter received data for CRC errror detection: ")

        if not input_check(check_input):
            print("Invalid input! Please enter a binary string.")
            continue

        check_result = crc_check(check_input, divisor_input, crc_value)
        print("CRC check result:", check_result)

        if check_result == '0' * len(crc_value):
            print("No error is detected.")
        else:
            print("Error is detected. CRC check result should be all zeroes for zero error.")
        print("-----------------------------------------------------------------------------------------------------------------------------------------------")
    elif choice == "2":
        rows = int(input("Enter number of rows: "))
        sender= []
        receiver=[]

        for i in range(rows):
            ip = input("Enter data: ")
            if not input_check(ip):
                print("Invalid input! Please enter a binary string.")
                break
            sender.append(ip)
        else:
            checksum_bits = 8 if len(sender[0]) <= 8 else 16
            checksum = calculate_checksum(sender)
            print(f"Checksum value ({checksum_bits}-bit):", checksum)
            
            for i in range(rows):
                ip = input("Enter received data for checksum error detection: ")
                if not input_check(ip):
                    print("Invalid input! Please enter a binary string.")
                    break
                receiver.append(ip)
            
            receiver_side_result = calculate_checksum(receiver + [checksum])
            print(f"Checksum value ({checksum_bits}-bit) at the receiver side:", receiver_side_result)

            if receiver_side_result == '0' * checksum_bits:
                print("No error is detected in the code.")
            else:
                print("Error is detected in the code.")
            print("-----------------------------------------------------------------------------------------------------------------------------------------------")

    elif choice == "3":
        print("Exiting...")
        print("-----------------------------------------------------------------------------------------------------------------------------------------------")
        break

    else:
        print("Invalid choice! Please enter a valid option.")